test/integration/protocol_integration_test.cc

```diff
@@ -128,6 +128,75 @@ TEST_P(DownstreamProtocolIntegrationTest, RouterRedirect) {
             response->headers().get(Http::Headers::get().Location)[0]->value().getStringView());
 }
 
+// Add a filter that overrides the cached route/cluster selection with a DelegatingRoute
+TEST_P(DownstreamProtocolIntegrationTest, RouterClusterFromDelegatingRoute) {
+  EXCLUDE_UPSTREAM_HTTP3;
+
+  config_helper_.addFilter(R"EOF(
+  name: set-route-filter
+  )EOF");
+
+  // Upstreams: cluster_0, cluster_override
+  setUpstreamCount(2);
+  config_helper_.addConfigModifier([](envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+    std::string cluster_yaml = R"EOF(
+            name: cluster_override
+            connect_timeout: 1.250s
+            type: ORIGINAL_DST
+            lb_policy: CLUSTER_PROVIDED
+            original_dst_lb_config:
+              use_http_header: true
+          )EOF";
+    envoy::config::cluster::v3::Cluster cluster_config;
+    TestUtility::loadFromYaml(cluster_yaml, cluster_config);
+    auto* orig_dst_cluster = bootstrap.mutable_static_resources()->add_clusters();
+    orig_dst_cluster->MergeFrom(cluster_config);
+  });
+
+  auto host_co =
+      config_helper_.createVirtualHost("cluster_override vhost", "/some/path", "cluster_override");
+  config_helper_.addVirtualHost(host_co);
+
+  auto host_foo = config_helper_.createVirtualHost("cluster_0 vhost", "/some/path", "cluster_0");
+  config_helper_.addVirtualHost(host_foo);
+
+  initialize();
+
+  std::string ip;
+  if (GetParam().version == Network::Address::IpVersion::v4) {
+    ip = "127.0.0.1";
+  } else {
+    ip = "[::1]";
+  }
+  const std::string ip_port_pair =
+      absl::StrCat(ip, ":", fake_upstreams_[0]->localAddress()->ip()->port());
```

Comment:
If we're gonna keep this to test the original dst cluster we should at least comment why we need all this additional logic in this test